### DISCUSSION TITLE
Compatible with PyInstaller?

### DISCUSSION THREAD

**Discussion by imba-tjd** [Kludex/uvicorn#1172] [Ideas]
I see uvicorn uses `importer.py` so that PyInstaller can't correctly freeze this lib by default.

PyInstaller has a `hook` feature to fix this kind of issue.

I think all we need to do is

1. in https://github.com/pyinstaller/pyinstaller-hooks-contrib/tree/master/src/_pyinstaller_hooks_contrib/hooks
2. create `hook-uvicorn.py`
3. with this content
```
# ------------------------------------------------------------------
# Copyright (c) 2020 PyInstaller Development Team.
#
# This file is distributed under the terms of the GNU General Public
# License (version 2.0 or later).
#
# The full license is available in LICENSE.GPL.txt, distributed with
# this software.
#
# SPDX-License-Identifier: GPL-2.0-or-later
# ------------------------------------------------------------------

from PyInstaller.utils.hooks import collect_submodules

hiddenimports = collect_submodules('uvicorn')
```
4. create a changelog in https://github.com/imba-tjd/pyinstaller-hooks-contrib/tree/master/news

The `collect_submodules('uvicorn')` basically means import everything in uvicorn.

FYI:

* https://pyinstaller.readthedocs.io/en/stable/hooks.html
* https://github.com/pyinstaller/pyinstaller/pull/4664

Any thoughts?

---

PR have been merged. https://github.com/pyinstaller/pyinstaller-hooks-contrib/pull/300
It will work by default after a release.

**Comment by Kludex** [comment]
What kind of thoughts are you looking for? Do we need to do something on uvicorn for that?

**Reply by imba-tjd** [reply]
General discussion. I'm new to both uvicorn and PyInstaller.
I was about to say "only specify what can't be auto detected but is required".
But after checking the auto-detecting result, it's very close to include everything. So I think that's it. I'll make a PR on PyInstaller side.

**Comment by graingert** [comment]
would it work better if uvicorn used `pkg_util.resolve_name`?

**Reply by imba-tjd** [reply]
Not likely. PyInstaller says they can't detect `__import__` and `sys.path` ways to use module. I think uvicorn doesn't need to do any change.

**Reply by graingert** [reply]
afaik it's only used to import your app from the command line via the `uvicorn ham.spam:create_app --factory` syntax and so pyinstaller won't know what that app will be

**Reply by imba-tjd** [reply]
Anyway. Without any interfere, PyInstaller won't include `lifespan`, `loops`, `protocols/http/auto.py`, `workers.py`, maybe etc.

**Reply by graingert** [reply]
Ah yes of course, some of those are optional based on what other deps are installed
